### PR TITLE
[WIP] Adventures in the amazing world of specification of dependencies and predictiveness

### PR DESCRIPTION
**Pull request by nulltoken**
## Pull request description
@linkdotnet Early work in the land of #108 

Not a real PR as of now, rather a test bed for expectations (and tests that we'd like to see passing).

### PR meta checklist
- [ ] Pull request is targeted at `main` branch for code   
- [ ] Pull request is linked to all related issues, if any.

### Code PR specific checklist
- [ ] My code follows the code style of this project and AspNetCore coding guidelines.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [ ] I have updated the appropriate sub section in the _CHANGELOG.md_.
- [ ] I have added, updated or removed tests to according to my changes.
  - [ ] All tests passed.

### CODE REVIEW

**Comment by nulltoken** [comment]
@linkdotnet Not all tests are passing.

Before going further with how the code should be changed to make them pass, I'd like some feedback from you about the following points:

- Do the new added tests make sense?
- What are new scenarios that we'd like to add?

Note: I haven't wandered in the region of dynamic jobs. Yet.
Before diving into that, I'd like to ensure "how the lib should behave with 'simpler' jobs" is defined.

**Comment by linkdotnet** [comment]
I will have a look at this at the end of the week - unfortunately I am a bit swamped with stuff.

File: src/NCronJob/Configuration/Builder/DependencyBuilder.cs [hunk 1]
```diff
@@ -20,10 +20,43 @@ public sealed class DependencyBuilder<TPrincipalJob>
...
     }
 
+    /// <summary>
+    /// Adds a job that runs after the principal job has finished with a given <paramref name="parameter"/>.
+    /// <param name="jobName">The name of the registered job to run.</param>
+    /// <param name="parameter">An optional parameter that is passed down as the <see cref="JobExecutionContext"/> to the job.</param>
+    /// </summary>
+    /// <remarks>
+    /// </remarks>
+    public DependencyBuilder<TPrincipalJob> RunJob(string jobName, object? parameter = null)
```
Comment (linkdotnet) [inline src/NCronJob/Configuration/Builder/DependencyBuilder.cs:44]: I see where you are driving at. Imagine the following registration:
```csharp
builder.Services.AddNCronJob(o => 
{
  // Here comes your new function into play
  o.AddJob<...>(w => w.WithCronExpression("* * * * *")).ExecuteWhen(s => s.RunJob("Name", null));

  o.AddJob<DependentJob>(o => o.WithCronExpression("* * * * *").WithName("Name"));
});
```

**Technically** the user registered a job with the given job name. The order inside a container registration shouldn't be important. 

And that is where your `useNCronJobAsync` proposal from a few days may come in handy.

Comment (linkdotnet) [reply src/NCronJob/Configuration/Builder/DependencyBuilder.cs:44]: But this is more of an exotic use case (at least in my mind).

File: src/NCronJob/Registry/JobRegistry.cs [hunk 2]
```diff
@@ -30,7 +45,7 @@ public void Add(JobDefinition jobDefinition)
         {
             throw new InvalidOperationException(
                 $"""
-                Job registration conflict for type: {jobDefinition.Type.Name} detected. Another job with the same type, parameters, or cron expression already exists.
+                Job registration conflict for type '{jobDefinition.Type.Name}' detected. Another job with the same type, parameters, or cron expression already exists.
```
Comment (linkdotnet) [inline src/NCronJob/Registry/JobRegistry.cs:48]: We might want to special case that part. If the job is `typeof(DynamicJobFactory)` aka Lambda expression, then this will be more confusing than helpful

File: src/NCronJob/Configuration/Builder/DependencyBuilder.cs (hunk 1 above)
Comment (nulltoken) [reply src/NCronJob/Configuration/Builder/DependencyBuilder.cs:44]: >But this is more of an exotic use case (at least in my mind).

I'm not sure. I've tried a spike where all dependencies were required to be registered before jobs leveraging them.
Although that works, that's an awful user experience.

> The order inside a container registration shouldn't be important.

I have an idea how to resolve this

File: src/NCronJob/Registry/JobRegistry.cs (hunk 2 above)
Comment (nulltoken) [reply src/NCronJob/Registry/JobRegistry.cs:48]: I know. I'm starting with the easy part. I'll tackle Dynamic jobs once we've straighten the general expectations on typed jobs. (through expected passing/failing tests)

File: tests/NCronJob.Tests/NCronJobIntegrationTests.cs [hunk 3]
```diff
@@ -144,6 +144,26 @@ public async Task InstantJobShouldPassDownParameter()
...
+            n.AddJob<ParameterJob>(s => s.WithCronExpression("* * 30 2 *"));
+            n.AddJob<ParameterJob>(s => s.WithCronExpression("* * 31 2 *"));
+        });
+
+        var provider = CreateServiceProvider();
+        await provider.GetRequiredService<IHostedService>().StartAsync(CancellationToken);
+
+        var instantJobRegistry = provider.GetRequiredService<IInstantJobRegistry>();
+
+        Action act = () => instantJobRegistry.ForceRunInstantJob<ParameterJob>("something", CancellationToken);
```
[the remark this answers is not in the document]
Comment (nulltoken) [reply tests/NCronJob.Tests/NCronJobIntegrationTests.cs:179]: Updated the expectation:

```
        ServiceCollection.AddNCronJob(n =>
        {
            n.AddJob<ParameterJob>(s => s.WithCronExpression("* * 30 2 *"));
            n.AddJob<ParameterJob>(s => s.WithCronExpression("* * 31 2 *"));
        });
 ```

should no longer be expected to throw.

However, I've added 

```
        ServiceCollection.AddNCronJob(n =>
        {
            n.AddJob<ParameterJob>(s => s.WithCronExpression("* * 30 2 *").WithParameter("13"));
            n.AddJob<ParameterJob>(s => s.WithCronExpression("* * 31 2 *"));
        });
```

which I believe should throw (as the behavior of the job may be altered by the parameter)

File: src/NCronJob/Registry/JobRegistry.cs (hunk 2 above)
Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs:48]: Hmmm I am not sure why this should throw.

In my mental model CRON jobs and instant jobs are two different things. The only thing they share is the execution logic but none of its metadata (well technically some of it like the retrypolicy stuff, dependent jobs and whether or not they can be executed, but nothing related to CRON jobs and its parameter).
Furthermore, users have the ability to add CRON jobs (`IRuntimeRegistry.TryRegister`) and parameters (`IRuntimeRegistry.UpdateParameter`)  or update schedules.

That might lead to a situation where 
1. The initial Instant job `A` works because there was no parameter specified
2. `IRuntimeRegistry.UpdateParameter`is called for `A` with a new parameter
3. Now the instant job fails with the given exception

The user would need to keep track of the parameters and needs to set it temporarily to something else or `null` so the instant trigger succeeds.

More and more we might want to check of we need distinct objects like `JobDefinition`, `JobRun` and `Trigger`.

That would solve some of our issues, mainly:
1. Dependent jobs can hang either on a Trigger or a JobDefinition (so they get triggered for instant jobs as well). The location where you call `ExecuteWhen` is key. But takes out ambiguity. **Maybe** we can even have distinct methods for that so it is painfully clear: That should be straightforward as can provide overloads to `AddJob`. For example if you "only" call: `AddJob().ExecuteWhen` it should always be on the type aka Jobdefinition, **but**: `AddJob(c => c.WithCronExpression())` should sit on the trigger.
2. Would make it also easier to distinct if we have duplicates. The JobDefinition should **always** be unique and we just have a bunch of triggers.

Sorry for the wall of text, but seeing that our system is so ambiguous tells me, we need to go in a different direction. Extending the current system may lead to more confusion

Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs:48]: A trigger might have a definition like that:
```csharp
/// <summary>
/// Base interface for all job triggers
/// </summary>
internal interface IJobTrigger
{
    bool ShouldFire(DateTimeOffset currentTime);
   
    DateTimeOffset? GetNextFireTime(DateTimeOffset after);
    
    object? Parameters { get; }
    
    JobPriority Priority { get; }
}

internal class CronTrigger : IJobTrigger 
{
    private readonly CronExpression cronExpression;
    private readonly TimeZoneInfo timeZone;
    
    public CronTrigger(string cronExpression, object? parameters = null, TimeZoneInfo? timeZone = null)
    {
        this.cronExpression = CronExpression.Parse(cronExpression);
        this.timeZone = timeZone ?? TimeZoneInfo.Utc;
        this.Parameters = parameters;
        this.Priority = JobPriority.Normal;
    }

    public object? Parameters { get; }
    public JobPriority Priority { get; init; }

    public bool ShouldFire(DateTimeOffset currentTime) 
    {
        var next = GetNextFireTime(currentTime.AddSeconds(-1));
        return next?.Equals(currentTime) ?? false;
    }

    public DateTimeOffset? GetNextFireTime(DateTimeOffset after)
        => cronExpression.GetNextOccurrence(after, timeZone); 
}

internal class InstantTrigger : IJobTrigger
{
    public InstantTrigger(object? parameters = null, DateTimeOffset? scheduledTime = null)
    {
        Parameters = parameters;
        ScheduledTime = scheduledTime ?? DateTimeOffset.UtcNow;
        Priority = JobPriority.High;
    }

    public DateTimeOffset ScheduledTime { get; }
    public object? Parameters { get; }
    public JobPriority Priority { get; init; }

    public bool ShouldFire(DateTimeOffset currentTime)
        => currentTime >= ScheduledTime;

    public DateTimeOffset? GetNextFireTime(DateTimeOffset after)
        => after <= ScheduledTime ? ScheduledTime : null;
}
```

The `JobDefintion` might then look like this:
```csharp
internal sealed class JobDefinition
{
    public JobDefinition(Type type, IJobTrigger trigger)
    {
        Type = type;
        Trigger = trigger;
    }

    public Type Type { get; }
    public IJobTrigger Trigger { get; }
    // The rest of our props here
}
```

Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs:48]: Also we have this beauty which might get solved:
```csharp
builder.Services.AddNCronJob(o => 
{
  o.AddJob<MyJob>().ExecuteWhen(s => s.RunJob<JobA>());
  o.AddJob<MyJob>(s => s.WithCronExpression("* * * * *")).ExecuteWhen(s => s.RunJob<JobB>());
```

If "* * * * *" is `true` - what dependent jobs should be executed? I'd argue `JobA` **and `JobB`. And this would easy to model with the `Trigger`. well "easier".

Comment (nulltoken) [reply src/NCronJob/Registry/JobRegistry.cs:48]: >Sorry for the wall of text, but seeing that our system is so ambiguous tells me, we need to go in a different direction. Extending the current system may lead to more confusion

That's exactly my current train of thought. This also explain why I took an "ambiguity nailing" approach.
I believe NCronJob needs a clear model so that it's easy to reason about (from both maintainer and consumer standpoints).

> Also we have this beauty which might get solved:
> 
> ```cs
> builder.Services.AddNCronJob(o => 
> {
>   o.AddJob<MyJob>().ExecuteWhen(s => s.RunJob<JobA>());
>   o.AddJob<MyJob>(s => s.WithCronExpression("* * * * *")).ExecuteWhen(s => s.RunJob<JobB>());
> }
> ```
> 
> If "* * * * *" is `true` - what dependent jobs should be executed? I'd argue `JobA` **and `JobB`. And this would easy to model with the `Trigger`. well "easier".

I agree with the statement 'I'd argue `JobA` and `JobB` should be triggered` (considering a clear separation between the definition and the triggering model).

However, and that's where I feel uneasy, this is hard to read.
It requires **your** brain to decipher this and foresee the execution process.

However, the piece of code above can be rewritten into
```cs
builder.Services.AddNCronJob(o => 
{
  o.AddJob<MyJob>(s => s.WithCronExpression("* * * * *")).ExecuteWhen(s => s.RunJob<JobA>().RunJob<JobB>());
}
```

Which should generate the same execution process while being (much?) easier to read.

My belief is that we should do our best to drive the user to write code that's easy to review and maintain.

Thoughts?

Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs:48]: I am 100% on your side.
My idea for `v4` would be to have better naming. Now I will overexaggerate here:
```csharp
builder.Services.AddNCronJob(o => 
{
  o.AddJob<MyJob>().ExecuteWhenJobTypeHasRun(s => s.RunJob<JobA>());
  o.AddJob<MyJob>(s => s.WithCronExpression("* * * * *")).ExecuteWhenSpecificJobWithParametersHasRun(s => s.RunJob<JobB>());
}
```

Having two different names may sort out some of the trouble.

The question is, what should we offer users? I do not like the approach of Quartz (I guess?), where the registration just returns a unique ID you can pass in everywhere. That is just not simple! While it takes out all the ambiguity, it is dreadful.

I tried to spike/poc something with Triggers but didn't have enough time to finish it. It helps in certain situations to reason about what's going on. But that is just because that is the only idea that came to my mind - I am absolutely open to entirely different approaches (as long as we keep the "essence" of the lib).

Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs:48]:
```csharp
builder.Services.AddNCronJob(o => 
{
  o.AddJob<MyJob>(s => s.WithCronExpression("* * * * *")).ExecuteWhen(s => s.RunJob<JobA>().RunJob<JobB>());
}
```

puuhhh - you dropping here something. Shall they be the same as the registration above?
I'd argue no :D as both are sitting on the cron expression thingy.

Comment (nulltoken) [reply src/NCronJob/Registry/JobRegistry.cs:48]: > In my mental model CRON jobs and instant jobs are two different things. The only thing they share is the execution logic but none of its metadata (well technically some of it like the retrypolicy stuff, dependent jobs and whether or not they can be executed, but nothing related to CRON jobs and its parameter).

@linkdotnet Hrmpf. I completely overlooked this paragraph. And some of it actually triggers a lot of questions in my mind.

I've never linked CRON expression with parameters. In my mind, those two things are completely orthogonal (FWIW, 
I've got a local branch with some code change slightly tweaking the fluent interface in order to allow a `WithParameter().WithName()` chain without a `WithCronExpression()` root.)

Reasoning (at least how I see it):
- Parameters are a way to optionally alter the behavior of how a job runs.
- Those parameters can be set through a `AddJob()` call
- They can be overridden through an instant job execution

Thoughts? Is there a world where parameters don't depend on Cron expressions?

Comment (nulltoken) [reply src/NCronJob/Registry/JobRegistry.cs:48]: Going a little down further this path, I "see" actually two different phases/steps (independently of how the API make them happen).

The Definition:
- The root job, its parametrization, the orchestration of dependent steps (with RunJob() allowing the optional override of the parameter). Those steps being either inline jobs or "references" to other root jobs.

The Scheduling:
- A way to trigger root jobs at regular intervals

All layers of new registrations contributing to the Definition should lead to a list of entries and references that are unambiguous. When an ambiguity is detected, the lib complains and the "Definition" should be updated to leverage "Names" to resolve those ambiguities (RunJob() may have to be updated to accept a jobName to make this happen and JobOptionBuilder might also need to expose `WithName()`).

Instant Jobs are only a special case of Scheduling

RuntimeRegistry ensures through the `TryRegister()` method that the Definition is kept unambiguous.

To sum up:
- Easy job setups would still be easy to define.
- Complex orchestrations involving reused jobs in different execution contexts would be possible in a safe way, the registration engine ensuring the absence of ambiguity at runtime

What's your feeling about this?

Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs:48]: > I've never linked CRON expression with parameters. In my mind, those two things are completely orthogonal (FWIW,

So right now that is the model. Defining a job: `AddJob<MyJob>(s => s.WithCronExpression("* * * * *").WithParameter("foo"))` is exactly that.

> I've got a local branch with some code change slightly tweaking the fluent interface in order to allow a WithParameter().WithName()

True that - it is currently only possible to register CRON jobs with a name. If we take the current world  only, then `WithParameter` without a CRON expression doesn't make sense, as you can specify parameters for instant job when calling them. 

> They can be overridden through an instant job execution

That is an fair assumption - and I like that mental model. The current API might make it a bit awkward. Imagine you register a Job with two CRON Expressions and each of them has different parameters. Now you trigger an instant job **without** passing in arguments - what will be used?

> Going a little down further this path, I "see" actually two different phases/steps 

It goes exactly into the direction you mentioned some time ago, where `AddNCronJob` is the collection phase and `Use` is the configuration/kickstart phase.

Comment (nulltoken) [reply src/NCronJob/Registry/JobRegistry.cs:48]: > That is an fair assumption - and I like that mental model. The current API might make it a bit awkward. Imagine you register a Job with two CRON Expressions and each of them has different parameters. Now you trigger an instant job **without** passing in arguments - what will be used?

This is actually where an ambiguity would be raised at runtime. In order to resolve it, one would have to identify a specific job through its name to trigger it.

Some flavors of ambiguities can be detected before runtime (during the registration) and would prevent the app from running.

Some others (as the one you just described) can be identified as well (and for instance logged as a warning (eg. "Job X is registered Y times without name. Trying to invoke it through an InstantJob would lead to a runtime exception.") but wouldn't prevent the app from starting as the lib cannot be 100% sure whether or not that InstantJob invocation will happen or not). 

>True that - it is currently only possible to register CRON jobs with a name. If we take the current world only, then WithParameter without a CRON expression doesn't make sense, as you can specify parameters for instant job when calling them.

There's a case when this could make sense even in the current world.

```
n.AddJob<Reporting>(s => s..WithParameter("using standard channels"));
n.AddJob<TaskOne>(s => s.WithCronExpression(Monthly)))
    .ExecuteWhen(s => s.RunJob<Reporting>());
n.AddJob<TaskTwo>(s => s.WithCronExpression(Daily)))
    .ExecuteWhen(s => s.RunJob<Reporting>());
n.AddJob<TaskThree>(s => s.WithCronExpression(Daily)))
    .ExecuteWhen(s => s.RunJob<Reporting>());
```

And then from an InstantJob `.ForceRunInstantJob<TaskOne>("using emergency channels");`

Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs:48]: By any chance, do you know how `Quartz`/`Hangfire` handles such situations?

> This is actually where an ambiguity would be raised at runtime. In order to resolve it, one would have to identify a specific job through its name to trigger it.

I fear that if we do this, we declassify Instant jobs to the extent that you always have to take care of naming, just because you want to trigger a job with different names. So you have to register the job twice at least - once for instant jobs, once for CRON (which could be also a route - offering a `AddInstantJob` method that operates on its own term).

The funny part is that I use this behavior in my own projects with **NCronJob**. I do have a Cron Job with Parameter "A" but the instant job triggers with Parameter "B" and I don't wanna think too much about this while building up the container.
